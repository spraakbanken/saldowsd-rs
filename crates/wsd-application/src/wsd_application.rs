use std::io;

use ::process_corpus::LemmaToken;
use process_corpus::process_corpus;
use saldo::SaldoLexicon;

pub trait WSDApplication {
    fn disambiguate_text(
        &self,
        text: Vec<Vec<LemmaToken>>,
    ) -> Vec<(Vec<LemmaToken>, Vec<Option<Vec<f32>>>)> {
        let mut out = Vec::new();
        for sen in text {
            out.push(self.disambiguate_sentence(sen));
        }
        out
    }
    fn disambiguate_sentence(
        &self,
        lts: Vec<LemmaToken>,
    ) -> (Vec<LemmaToken>, Vec<Option<Vec<f32>>>) {
        let mut disamb = Vec::new();
        for i in 0..lts.len() {
            disamb.push(self.disambiguate(&lts, i));
        }
        (lts, disamb)
    }
    fn disambiguate(&self, lts: &[LemmaToken], i: usize) -> Option<Vec<f32>>;
}

pub type SharedWSDApplication = Box<dyn WSDApplication>;

pub fn evaluate(_wsd: SharedWSDApplication, _eval_lemmas_file: &str, _eval_key_file: &str) {
    todo!("evaluate is not yet supported")
}

pub fn read_sentences(
    reader: &mut dyn io::BufRead,
    _saldo: Option<&SaldoLexicon>,
    sbxml: bool,
    batch_size: usize,
    _split_mwes: bool,
    _split_compounds: bool,
) -> io::Result<Vec<Vec<LemmaToken>>> {
    let mut out = Vec::new();
    while out.len() < batch_size {
        if sbxml {
            todo!("sbxml format is not yet supported")
        } else {
            let lts = match process_corpus::read_lemma_tokens(reader)? {
                Some(lts) => lts,
                None => return Ok(out),
            };
            out.push(lts);
        }
    }
    Ok(out)
}
