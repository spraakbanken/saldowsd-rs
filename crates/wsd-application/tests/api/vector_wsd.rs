use std::{fs, io};

use rstest::{fixture, rstest};

use wsd_application::{
    make_wsd_application,
    wsd_application::{disambiguate_sentences, DisambiguateOptions},
    SharedWSDApplication, TabFormat,
};

#[fixture]
fn sense_model() -> &'static str {
    "assets/sparv-wsd/models/scouse/ALL_512_128_w10_A2_140403_ctx1.bin"
}

#[fixture]
fn context_model() -> &'static str {
    "assets/sparv-wsd/models/scouse/lem_cbow0_s512_w10_NEW2_ctx.bin"
}
#[fixture]
fn vector_wsd(sense_model: &str, context_model: &str) -> SharedWSDApplication {
    let argv = &[
        // "-format=tab".to_string(),
        format!("-svFile={}", sense_model),
        format!("-cvFile={}", context_model),
        "-s1Prior=1".into(),
        "-decay=true".into(),
        "-contextWidth=10".into(),
        "-verbose=false".into(),
    ];
    make_wsd_application(None, "se.gu.spraakbanken.wsd.VectorWSD", argv).expect("VectorWSD created")
}

#[rstest]
fn test_vector_wsd(vector_wsd: SharedWSDApplication) -> eyre::Result<()> {
    let mut reader = io::BufReader::new(fs::File::open("assets/testing/example1.in.txt")?);
    let mut out = Vec::new();
    disambiguate_sentences(
        vector_wsd,
        &mut reader,
        &mut out,
        &TabFormat::default(),
        DisambiguateOptions::default(),
    )?;

    let actual = String::from_utf8(out)?;
    let actual: Vec<&str> = actual.split('\n').collect();
    insta::assert_debug_snapshot!(actual);
    Ok(())
}
