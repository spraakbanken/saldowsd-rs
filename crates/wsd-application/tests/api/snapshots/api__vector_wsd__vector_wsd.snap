---
source: crates/wsd-application/tests/api/vector_wsd.rs
expression: actual
---
[
    "_\t_\t_\t_\tArbetsmarknadsutskottet..nn\t_\t_",
    "_\t_\t_\t_\ts..pm\t_\t_",
    "_\t_\t_\t_\tbetänkande..nn\tbetänkande..1\t_",
    "_\t_\t_\t_\t2018..rg\t_\t_",
    "_\t_\t_\t_\t/..mid\t_\t_",
    "_\t_\t_\t_\t19..rg\t_\t_",
    "_\t_\t_\t_\t:..mid\t_\t_",
    "_\t_\t_\t_\tAU1..pm\t_\t_",
    "_\t_\t_\t_\tUtgiftsområde..nn\t_\t_",
    "_\t_\t_\t_\t13..rg\t_\t_",
    "_\t_\t_\t_\tjämställdhet..nn\tjämställdhet..1\t_",
    "_\t_\t_\t_\toch..kn\toch..1\t_",
    "_\t_\t_\t_\tnyanländ..av\tnyanländ..1\t_",
    "_\t_\t_\t_\tinvandrare..nn\tinvandrare..1\t_",
    "_\t_\t_\t_\tetablering..nn\tetablering..1\t_",
    "_\t_\t_\t_\tsammanfattning..nn\tsammanfattning..1\t_",
    "_\t_\t_\t_\tutskott..nn\tutskott..1|utskott..2|utskott..3\t0.0007421329|0.98321337|0.016044592",
    "_\t_\t_\t_\tföreslå..vb\tföreslå..1\t_",
    "_\t_\t_\t_\tatt..sn\tatt..1\t_",
    "_\t_\t_\t_\triksdag..nn\triksdag..1\t_",
    "_\t_\t_\t_\tanvisa..vb\tanvisa..1|anvisa..2\t0.9851393|0.014860547",
    "_\t_\t_\t_\tanslag..nn\tanslag..1|anslag..2|anslag..3\t0.9993586|0.00015822364|0.0004832597",
    "_\t_\t_\t_\ten..al\tden..1|en..2\t_",
    "_\t_\t_\t_\tför..pp\tför..1|för..5|för..6|för..7|för..9\t_",
    "_\t_\t_\t_\t2019..rg\t_\t_",
    "_\t_\t_\t_\toch..kn\toch..1\t_",
    "_\t_\t_\t_\tbemyndiga..vb\tbemyndiga..1\t_",
    "_\t_\t_\t_\tregering..nn\tregering..1\t_",
    "_\t_\t_\t_\tatt..sn\tatt..1\t_",
    "_\t_\t_\t_\tingå..vb\tingå..1|ingå..2\t0.94996685|0.05003296",
    "_\t_\t_\t_\tekonomisk..av\tekonomisk..1|ekonomisk..2|ekonomisk..3\t0.9764713|0.0030415813|0.020487025",
    "_\t_\t_\t_\tåtagande..nn\tåtagande..1\t_",
    "_\t_\t_\t_\tinom..pp\tinom..1\t_",
    "_\t_\t_\t_\tutgiftsområde..nn\t_\t_",
    "_\t_\t_\t_\t1..rg\t_\t_",
    "_\t_\t_\t_\t3..rg\t_\t_",
    "_\t_\t_\t_\tjämställdhet..nn\tjämställdhet..1\t_",
    "_\t_\t_\t_\toch..kn\toch..1\t_",
    "_\t_\t_\t_\tnyanländ..av\tnyanländ..1\t_",
    "_\t_\t_\t_\tinvandrare..nn\tinvandrare..1\t_",
    "_\t_\t_\t_\tetablering..nn\tetablering..1\t_",
    "_\t_\t_\t_\ti..pp\ti..2\t_",
    "_\t_\t_\t_\t_\t_\t_",
    "_\t_\t_\t_\tmed..pp\tmed..1|med..2\t_",
    "_\t_\t_\t_\ten..al\tden..1|en..2\t_",
    "_\t_\t_\t_\tförslag_2..nn|förslag..nn\tförslag..1|förslag..2\t1|0",
    "_\t_\t_\t_\tsom..hp\t_\t_",
    "_\t_\t_\t_\tlägga..vb\tlägga..1|lägga..2|lägga..3\t0.86679447|0.12821661|0.004989061",
    "_\t_\t_\t_\tfram..ab\tfram..1|fram..2|fram..3\t0.066161424|0.9018107|0.03202789",
    "_\t_\t_\t_\tgemensam..av\tgemensam..1\t_",
    "_\t_\t_\t_\tav..pp\tav..1\t_",
    "_\t_\t_\t_\tmoderat..nn\tmoderat..2\t_",
    "_\t_\t_\t_\toch..kn\toch..1\t_",
    "_\t_\t_\t_\tkristdemokrat..nn\tkristdemokrat..1\t_",
    "_\t_\t_\t_\t...mad\t_\t_",
    "_\t_\t_\t_\t$SENT$\t_\t_",
    "_\t_\t_\t_\ten..al\tden..1|en..2\t_",
    "_\t_\t_\t_\tföreslå..vb\tföreslå..1\t_",
    "_\t_\t_\t_\tanslag..nn\tanslag..1|anslag..2|anslag..3\t0.9987727|0.0005153681|0.00071193546",
    "_\t_\t_\t_\tuppgå..vb\tuppgå..1|uppgå..2\t0.99813825|0.0018617423",
    "_\t_\t_\t_\ttill..pp\ttill..1\t_",
    "_\t_\t_\t_\tsammanlagd..av\tsammanlagd..1\t_",
    "_\t_\t_\t_\tca..aba\tcirka..1\t_",
    "_\t_\t_\t_\t18..rg\t_\t_",
    "_\t_\t_\t_\tmiljard..nn\tmiljard..1\t_",
    "_\t_\t_\t_\tkrona..nn\tkrona..1|krona..2|krona..3\t0.9988465|0.00055660156|0.00059682806",
    "_\t_\t_\t_\tför..pp\tför..1|för..5|för..6|för..7|för..9\t_",
    "_\t_\t_\t_\t2019..rg\t_\t_",
    "_\t_\t_\t_\t...mad\t_\t_",
    "_\t_\t_\t_\t$SENT$\t_\t_",
    "_\t_\t_\t_\tutskott..nn\tutskott..1|utskott..2|utskott..3\t0.06320306|0.7548209|0.18197611",
    "_\t_\t_\t_\tförslag_2..nn|förslag..nn\tförslag..1|förslag..2\t1|0",
    "_\t_\t_\t_\tavvika..vb\tavvika..1|avvika..2\t0.5912454|0.40875474",
    "_\t_\t_\t_\tfrån..pp\tfrån..1\t_",
    "_\t_\t_\t_\tbudgetproposition..nn\tbudgetproposition..1\t_",
    "_\t_\t_\t_\tnär..ab\tnär..1|nära..1\t0.9939288|0.006071073",
    "_\t_\t_\t_\tden..pn\tden..2\t_",
    "_\t_\t_\t_\tgälla..vb\tgälla..1|gälla..2|gälla..4|gälla..3\t0.91496444|0.077562764|0.0042356816|0.0032370193",
    "_\t_\t_\t_\tanslag..nn\tanslag..1|anslag..2|anslag..3\t0.98988485|0.0038936287|0.0062217633",
    "_\t_\t_\t_\tEtableringsåtgärder..nn\t_\t_",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tKommunersättningar..nn\t_\t_",
    "_\t_\t_\t_\tvid..pp\tvid..1\t_",
    "_\t_\t_\t_\tflyktingmottagande..nn\tflyktingmottagande..1\t_",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tHemutrustningslån..nn\t_\t_",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tsärskilja..vb|särskild..av\tsärskild..1|särskilja..1\t0.4431396|0.55686045",
    "_\t_\t_\t_\tjämställdhetsåtgärder..nn\t_\t_",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tjämställdhet..nn\tjämställdhet..1\t_",
    "_\t_\t_\t_\tmyndighet..nn\tmyndighet..1|myndighet..2|myndighet..3\t0.9961731|0.0016595483|0.002167359",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tåtgärd..nn\tåtgärd..1\t_",
    "_\t_\t_\t_\tmot..pp\tmot..1|mot..2\t_",
    "_\t_\t_\t_\tsegregation..nn\tsegregation..1\t_",
    "_\t_\t_\t_\toch..kn\toch..1\t_",
    "_\t_\t_\t_\tdelegation..nn\tdelegation..1|delegation..2\t0.4626635|0.53733665",
    "_\t_\t_\t_\tmot..pp\tmot..1|mot..2\t_",
    "_\t_\t_\t_\tsegregation..nn\tsegregation..1\t_",
    "_\t_\t_\t_\tsamt..kn\tsamt..1\t_",
    "_\t_\t_\t_\tnär..ab\tnär..1|nära..1\t0.99522305|0.0047770976",
    "_\t_\t_\t_\tden..pn\tden..2\t_",
    "_\t_\t_\t_\tgälla..vb\tgälla..1|gälla..2|gälla..4|gälla..3\t0.9302709|0.066495255|0.001619833|0.0016139766",
    "_\t_\t_\t_\ten..al\tden..1|en..2\t_",
    "_\t_\t_\t_\tbeställning..nn\tbeställning..1\t_",
    "_\t_\t_\t_\tbemyndigande..nn\tbemyndigande..1\t_",
    "_\t_\t_\t_\t...mad\t_\t_",
    "_\t_\t_\t_\t$SENT$\t_\t_",
    "_\t_\t_\t_\ti..pp\ti..2\t_",
    "_\t_\t_\t_\tfråga..nn\tfråga..2|fråga..3\t0.90952575|0.09047447",
    "_\t_\t_\t_\tom..pp\tom..1|om..5\t_",
    "_\t_\t_\t_\tövrig..pn\tövrig..1\t_",
    "_\t_\t_\t_\tanslag..nn\tanslag..1|anslag..2|anslag..3\t0.9925965|0.003730983|0.0036726526",
    "_\t_\t_\t_\toch..kn\toch..1\t_",
    "_\t_\t_\t_\tbeställning..nn\tbeställning..1\t_",
    "_\t_\t_\t_\tbemyndigande..nn\tbemyndigande..1\t_",
    "_\t_\t_\t_\tinom..pp\tinom..1\t_",
    "_\t_\t_\t_\tutgiftsområdet..nn\t_\t_",
    "_\t_\t_\t_\tfinnas..vb|finna..vb\tfinna..1|finna..2|finnas..1\t0.54719967|0.006093521|0.44670677",
    "_\t_\t_\t_\tden..pn\tden..2\t_",
    "_\t_\t_\t_\tingen..pn\tingen..1\t_",
    "_\t_\t_\t_\tavvikelse..nn\tavvikelse..1|avvikelse..2\t0.97827023|0.021729657",
    "_\t_\t_\t_\tmellan..pp\tmellan..1\t_",
    "_\t_\t_\t_\tutskott..nn\tutskott..1|utskott..2|utskott..3\t0.055942018|0.83461386|0.10944395",
    "_\t_\t_\t_\tförslag_2..nn|förslag..nn\tförslag..1|förslag..2\t1|0",
    "_\t_\t_\t_\toch..kn\toch..1\t_",
    "_\t_\t_\t_\tförslag_2..nn|förslag..nn\tförslag..1|förslag..2\t1|0",
    "_\t_\t_\t_\ti..pp\ti..2\t_",
    "_\t_\t_\t_\tbudgetproposition..nn\tbudgetproposition..1\t_",
    "_\t_\t_\t_\t...mad\t_\t_",
    "_\t_\t_\t_\t$SENT$\t_\t_",
    "_\t_\t_\t_\tdärmed..ab\tdärmed..1\t_",
    "_\t_\t_\t_\tavstyrka..vb\tavstyrka..1\t_",
    "_\t_\t_\t_\tutskott..nn\tutskott..1|utskott..2|utskott..3\t0.0009681393|0.99808353|0.0009483777",
    "_\t_\t_\t_\tSverige..pm\tSverige..1\t_",
    "_\t_\t_\t_\tdemokrat..nn\tdemokrat..1\t_",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tCenterpartiet..pm\tCenterpartiet..1\t_",
    "_\t_\t_\t_\toch..kn\toch..1\t_",
    "_\t_\t_\t_\tliberal..nn\tliberal..2\t_",
    "_\t_\t_\t_\tbudgetförslag..nn\tbudgetförslag..1\t_",
    "_\t_\t_\t_\t...mad\t_\t_",
    "_\t_\t_\t_\t$SENT$\t_\t_",
    "_\t_\t_\t_\ti..pp\ti..2\t_",
    "_\t_\t_\t_\tbetänkande..nn\tbetänkande..1\t_",
    "_\t_\t_\t_\tfinnas..vb|finna..vb\tfinna..1|finna..2|finnas..1\t0.59131926|0.061062727|0.34761834",
    "_\t_\t_\t_\ten..al\tden..1|en..2\t_",
    "_\t_\t_\t_\treservation..nn\treservation..1\t_",
    "_\t_\t_\t_\t(..pad\t_\t_",
    "_\t_\t_\t_\tSD..pma\tSverigedemokraterna..1\t_",
    "_\t_\t_\t_\t)..pad\t_\t_",
    "_\t_\t_\t_\toch..kn\toch..1\t_",
    "_\t_\t_\t_\tfyra..nl\tfyra..1\t_",
    "_\t_\t_\t_\tsärskilja..vb|särskild..av\tsärskild..1|särskilja..1\t0.41285187|0.58714855",
    "_\t_\t_\t_\tyttrande..nn\tyttrande..1\t_",
    "_\t_\t_\t_\t(..pad\t_\t_",
    "_\t_\t_\t_\tS..nna|s..nna|s..nn\tsvavel..1|s..1|sida..2\t0.2782383|0.7186184|0.0031431424",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tSD..pma\tSverigedemokraterna..1\t_",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tC..pm\t_\t_",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tV..pm\t_\t_",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tL..pm\t_\t_",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tMP..pm\t_\t_",
    "_\t_\t_\t_\t)..pad\t_\t_",
    "_\t_\t_\t_\t...mad\t_\t_",
    "_\t_\t_\t_\t$SENT$\t_\t_",
    "_\t_\t_\t_\tledamot..nn\tledamot..1\t_",
    "_\t_\t_\t_\tfrån..pp\tfrån..1\t_",
    "_\t_\t_\t_\tsocialdemokrat..nn\tsocialdemokrat..1\t_",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tcenterparti..nn\tCenterpartiet..1|centerparti..1\t0.40733868|0.5926613",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tvänster..nn\tvänster..4|vänster..2|vänster..3\t0.000029189712|0.99994063|0.000030163777",
    "_\t_\t_\t_\tparti..nn\tparti..1|parti..2|parti..3|parti..4|parti..5\t0.9996558|0.000030636595|0.000042675896|0.00008039402|0.00019040202",
    "_\t_\t_\t_\t,..mid\t_\t_",
    "_\t_\t_\t_\tliberal..nn\tliberal..2\t_",
    "_\t_\t_\t_\toch..kn\toch..1\t_",
    "_\t_\t_\t_\tMiljöpartiet..nn\t_\t_",
    "_\t_\t_\t_\tavstå..vb\tavstå..1|avstå..2\t0.9967124|0.0032876313",
    "_\t_\t_\t_\tfrån..pp\tfrån..1\t_",
    "_\t_\t_\t_\tställningstagande..nn\tställningstagande..1\t_",
    "_\t_\t_\t_\toch..kn\toch..1\t_",
    "_\t_\t_\t_\tredovisa..vb\tredovisa..1\t_",
    "_\t_\t_\t_\ti..pp\ti..2\t_",
    "_\t_\t_\t_\tställ..nn|ställe..nn\tställ..1|ställ..2|ställe..1\t0.035026606|0.4990451|0.46592838",
    "_\t_\t_\t_\tsig..pn\tsig..1\t_",
    "_\t_\t_\t_\tövervägande..nn\tövervägande..1|övervägande..3\t0.9971949|0.0028052244",
    "_\t_\t_\t_\ti..pp\ti..2\t_",
    "_\t_\t_\t_\tsärskilja..vb|särskild..av\tsärskild..1|särskilja..1\t0.5400561|0.45994383",
    "_\t_\t_\t_\tyttrande..nn\tyttrande..1\t_",
    "_\t_\t_\t_\t...mad\t_\t_",
    "_\t_\t_\t_\t$SENT$\t_\t_",
    "_\t_\t_\t_\tutskott..nn\tutskott..1|utskott..2|utskott..3\t0.0022177454|0.9844655|0.013316945",
    "_\t_\t_\t_\tföreslå..vb\tföreslå..1\t_",
    "_\t_\t_\t_\tatt..sn\tatt..1\t_",
    "_\t_\t_\t_\tärende..nn\tärende..1|ärende..2\t0.08174549|0.91825473",
    "_\t_\t_\t_\tavgöra..vb\tavgöra..1\t_",
    "_\t_\t_\t_\ttrots..pp\ttrots..1\t_",
    "_\t_\t_\t_\tatt..sn\tatt..1\t_",
    "_\t_\t_\t_\tden..pn\tden..2\t_",
    "_\t_\t_\t_\tha..vb\tha..1|ha..3\t0.98158973|0.018410355",
    "_\t_\t_\t_\tvara..vb\tvara..1\t_",
    "_\t_\t_\t_\ttillgänglig..av\ttillgänglig..1\t_",
    "_\t_\t_\t_\tkort..av\tkort..1\t_",
    "_\t_\t_\t_\ttid..nn\ttid..1\t_",
    "_\t_\t_\t_\tän..kn\t_\t_",
    "_\t_\t_\t_\ttvå..nl\ttvå..1\t_",
    "_\t_\t_\t_\tvardag..nn\tvardag..1\t_",
    "_\t_\t_\t_\tföre..pp\tföre..1\t_",
    "_\t_\t_\t_\ten..al|den..pn\tden..2|den..1|en..2\t_",
    "_\t_\t_\t_\tdag..nn\tdag..1|dag..2\t0.8698742|0.13012566",
    "_\t_\t_\t_\tdå..ab\tdå..1|då..3\t0.806499|0.19350107",
    "_\t_\t_\t_\tden..pn\tden..2\t_",
    "_\t_\t_\t_\tbehandla..vb\tbehandla..1|behandla..2|behandla..3\t0.8148463|0.10415241|0.08100135",
    "_\t_\t_\t_\t...mad\t_\t_",
    "_\t_\t_\t_\t$SENT$\t_\t_",
    "_\t_\t_\t_\tbehandla..vb\tbehandla..1|behandla..2|behandla..3\t0.8177202|0.11004514|0.072234474",
    "_\t_\t_\t_\tförslag_2..nn|förslag..nn\tförslag..1|förslag..2\t1|0",
    "_\t_\t_\t_\t$SENT$\t_\t_",
    "",
    "",
]
